sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala

```diff
@@ -263,6 +267,127 @@ case class GpuStringTrimRight(column: Expression, trimParameters: Option[Express
+    StringType +: Seq.fill(children.size - 1)(arrayOrStr)
+  }
+
+  private def concatArrayCol( sep: Either[GpuScalar, GpuColumnVector],
+      cv: ColumnView): GpuColumnVector = {
+    sep match {
+      case Left(sepScalar) =>
+        withResource(GpuScalar.from("", StringType)) { emptyStrScalar =>
+          GpuColumnVector.from(cv.stringConcatenateListElements(sepScalar.getBase, emptyStrScalar,
+            false, false), dataType)
+        }
+      case Right(sepVec) =>
+        GpuColumnVector.from(cv.stringConcatenateListElements(sepVec.getBase), dataType)
+    }
+  }
+
+  private def processSingleColScalarSep(cv: ColumnVector): GpuColumnVector = {
+    // single column with scalar separator just replace any nulls with empty string
+    withResource(Scalar.fromString("")) { emptyStrScalar =>
+      GpuColumnVector.from(cv.replaceNulls(emptyStrScalar), dataType)
+    }
+  }
+
+  private def stringConcatSeparatorScalar(columns: ArrayBuffer[ColumnVector],
+      sep: GpuScalar): GpuColumnVector = {
+    withResource(Scalar.fromString("")) { emptyStrScalar =>
+      GpuColumnVector.from(ColumnVector.stringConcatenate(sep.getBase, emptyStrScalar,
+        columns.toArray[ColumnView], false), dataType)
+    }
+  }
+
+  private def stringConcatSeparatorVector(columns: ArrayBuffer[ColumnVector],
+      sep: GpuColumnVector): GpuColumnVector = {
+    // GpuOverrides doesn't allow only specifying a separator, you have to specify at
+    // least one value column
+    GpuColumnVector.from(ColumnVector.stringConcatenate(columns.toArray[ColumnView],
+      sep.getBase), dataType)
+  }
+
+  private def resolveColumnVectorAndConcatArrayCols(
+      expr: Expression,
+      numRows: Int,
+      sep: Either[GpuScalar, GpuColumnVector],
+      batch: ColumnarBatch): GpuColumnVector = {
+    withResourceIfAllowed(expr.columnarEval(batch)) {
+      case vector: GpuColumnVector =>
+        vector.dataType() match {
+          case ArrayType(st: StringType, nullable) => concatArrayCol(sep, vector.getBase)
+          case _ => vector.incRefCount()
+        }
+      case s: GpuScalar =>
+        s.dataType match {
+          case ArrayType(st: StringType, nullable) =>
+            // we have to first concatenate any array types
+            withResource(GpuColumnVector.from(s, numRows, s.dataType).getBase) { cv =>
+              concatArrayCol(sep, cv)
+            }
+          case _ => GpuColumnVector.from(s, numRows, s.dataType)
+        }
+      case other =>
+        throw new IllegalArgumentException(s"Cannot resolve a ColumnVector from the value:" +
+          s" $other. Please convert it to a GpuScalar or a GpuColumnVector before returning.")
+      }
+  }
+
+  override def columnarEval(batch: ColumnarBatch): Any = {
+    val numRows = batch.numRows()
+    var sep: Either[GpuScalar, GpuColumnVector] = null
+    try {
+      // get the separator and check for null scalar separator
+      val sepScalarNullOpt = children.head.columnarEval(batch) match {
```

Comment:
This block returns `sepScalarNullOpt` but also sets the mutable `sep` as a side-effect. Could we just return a tuple of `(sep, sepScalarNullOpt)` instead?

Reply:
so originally I had this as separate function but that has potential to leak the sep.  as far as I know you can't assign an existing var to tuple returned, if you know a way let me know.  If we don't set sep directly then there is a chance it leaks as the finally wouldn't catch it.  its very small but was just trying to set such that finally would be able to clean up.

Reply:
I played with a couple of ideas but they turned out to be more trouble than they were worth